src/main/java/org/apache/commons/io/FileSystem.java

```diff
@@ -624,75 +631,102 @@ CharSequence trimExtension(final CharSequence cs) {
+                final CoderResult cr = encoder.encode(in, out, true);
+
+                if (cr.isUnderflow()) {
+                    // Entire candidate fit within maxFileNameLength bytes.
+                    return value;
+                }
 
-        // Slow path: encode into a fixed-size byte buffer.
-        final ByteBuffer out = ByteBuffer.allocate(maxBytes);
-        final CharBuffer in = CharBuffer.wrap(candidate);
+                // We ran out of space mid-encode: truncate BEFORE the offending character.
+                return value.subSequence(0, in.position());
+            }
+        },
 
-        // Encode until the first character that would exceed the byte budget.
-        final CoderResult cr = encoder.encode(in, out, true);
+        /** Length measured as UTF-16 code units (i.e., {@code CharSequence.length()}). */
+        UTF16_CHARS {
+            @Override
+            int getLength(final CharSequence value, final Charset charset) {
+                return value.length();
+            }
 
-        if (cr.isUnderflow()) {
-            // Entire candidate fit within maxFileNameLength bytes.
-            return candidate;
+            @Override
+            CharSequence truncate(final CharSequence value, final int limit, final Charset charset) {
+                return value.length() <= limit ? value : value.subSequence(0, limit);
+            }
+        };
+
+        /**
+         * Gets the measured length in this strategy’s unit.
+         *
+         * @param value The value to measure, not null.
+         * @param charset The charset to use when measuring in bytes.
+         * @return The length in this strategy’s unit.
+         */
+        abstract int getLength(CharSequence value, Charset charset);
+
+        /**
+         * Tests if the measured length is less or equal the {@code limit}.
+         *
+         * @param value The value to measure, not null.
+         * @param limit The limit to compare to.
+         * @param charset The charset to use when measuring in bytes.
+         * @return {@code true} if the measured length is less or equal the {@code limit}, {@code false} otherwise.
+         */
+        final boolean isWithinLimit(final CharSequence value, final int limit, final Charset charset) {
+            return getLength(value, charset) <= limit;
         }
 
-        // We ran out of space mid-encode: truncate BEFORE the offending character.
-        return candidate.subSequence(0, in.position());
+        /**
+         * Truncates to {@code limit} in this strategy’s unit (no-op if already within limit).
+         *
+         * @param value The value to truncate, not null.
+         * @param limit The limit to truncate to.
+         * @param charset The charset to use when measuring in bytes.
+         * @return The truncated value, not null.
+         */
+        abstract CharSequence truncate(CharSequence value, int limit, Charset charset);
```

Comment:
Should truncate at least be able to preserve extensions? Also make sure that after truncation file name is still valid. And can truncation also affect parents? Should there be a minimum length the remaining name has (don’t truncate to a single char, etc)

Reply:
Fixed in https://github.com/apache/commons-io/pull/781/commits/a0b99bfdb01ce7cc218475100f33ff182a37a8ed.